// @ngneat/spectator/esm2020/lib/spectator-routing/initial-module.mjs
import { ActivatedRoute, Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { Subject } from 'rxjs';
import { initialSpectatorModule } from '../spectator/initial-module';
import { ActivatedRouteStub } from './activated-route-stub';
import { RouterStub } from './router-stub';
/**
 * @internal
 */
export function initialRoutingModule(options) {
    const moduleMetadata = initialSpectatorModule(options);
    if (options.stubsEnabled) {
        moduleMetadata.imports.push(RouterTestingModule);
        moduleMetadata.providers.push(options.mockProvider(RouterStub, {
            events: new Subject(),
            emitRouterEvent(event) {
                this.events.next(event);
            },
            serializeUrl() {
                return '/';
            },
        }), {
            provide: Router,
            useExisting: RouterStub,
        });
        moduleMetadata.providers.push({
            provide: ActivatedRouteStub,
            useValue: new ActivatedRouteStub({
                params: options.params,
                queryParams: options.queryParams,
                data: options.data,
            }),
        }, {
            provide: ActivatedRoute,
            useExisting: ActivatedRouteStub,
        });
    }
    else {
        moduleMetadata.imports.push(RouterTestingModule.withRoutes(options.routes));
    }
    return moduleMetadata;
}
