// @ngneat/spectator/esm2020/lib/dom-selectors.mjs
import { queries as DOMQueries, getDefaultNormalizer } from '@testing-library/dom';
export class DOMSelector {
    // Wrap selector functions in a class to make reflection easier in getChild
    constructor(execute) {
        this.execute = execute;
    }
}
export const byLabel = (matcher, options) => new DOMSelector(el => DOMQueries.queryAllByLabelText(el, matcher, options));
export const byPlaceholder = (matcher, options) => new DOMSelector(el => DOMQueries.queryAllByPlaceholderText(el, matcher, options));
export const byText = (matcher, options) => new DOMSelector(el => DOMQueries.queryAllByText(el, matcher, options));
export const byTextContent = (matcher, options) => {
    let textContentMatcher;
    const normalizer = options?.normalizer || getDefaultNormalizer(options);
    const getTextContent = (elem) => normalizer(elem?.textContent ?? '');
    if (typeof matcher === 'string' || typeof matcher === 'number') {
        textContentMatcher = (_, elem) => {
            if (options?.exact === false) {
                return (getTextContent(elem)
                    .toLowerCase()
                    .indexOf(matcher.toString().toLowerCase()) >= 0);
            }
            return getTextContent(elem) === matcher.toString();
        };
    }
    else if (matcher instanceof RegExp) {
        textContentMatcher = (_, elem) => matcher.test(getTextContent(elem));
    }
    else if (typeof matcher === 'function') {
        textContentMatcher = (_, elem) => matcher(getTextContent(elem), elem);
    }
    else {
        throw new Error(`Matcher type not supported: ${typeof matcher}`);
    }
    return new DOMSelector(el => DOMQueries.queryAllByText(el, textContentMatcher, options));
};
export const byAltText = (matcher, options) => new DOMSelector(el => DOMQueries.queryAllByAltText(el, matcher, options));
export const byTitle = (matcher, options) => new DOMSelector(el => DOMQueries.queryAllByTitle(el, matcher, options));
export const byTestId = (matcher, options) => new DOMSelector(el => DOMQueries.queryAllByTestId(el, matcher, options));
export const byValue = (matcher, options) => new DOMSelector(el => DOMQueries.queryAllByDisplayValue(el, matcher, options));
export const byRole = (matcher, options) => new DOMSelector(el => DOMQueries.queryAllByRole(el, matcher, options));
