// @ngneat/spectator/esm2020/lib/base/initial-module.mjs
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { getGlobalsInjections } from '../globals-injections';
/**
 * @internal
 */
export function initialModule(options) {
    const globals = { imports: [], declarations: [], providers: [], ...getGlobalsInjections() };
    return {
        declarations: [...globals.declarations, ...options.declarations, ...options.entryComponents],
        imports: [...(options.disableAnimations ? [NoopAnimationsModule] : []), ...globals.imports, ...options.imports],
        providers: [...globals.providers, ...options.providers, ...options.mocks.map((type) => options.mockProvider(type))],
        entryComponents: [...options.entryComponents],
        teardown: 
        // Caretaker note: we don't want to merge the `globals.teardown` and `options.teardown`, since `options.teardown`
        // is always defined. If the user calls `defineGlobalsInjections({ teardown: { ... } })` and we merge it with
        // `options.teardown`, then `options.teardown` will always override global options.
        { ...(globals.teardown || options.teardown) },
    };
}
