// @ngneat/spectator/esm2020/lib/spectator-host/host-component.mjs
import { Component, NgModule } from '@angular/core';
import * as i0 from "@angular/core";
export class HostComponent {
}
HostComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: HostComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
HostComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.0.2", type: HostComponent, selector: "ng-component", ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: HostComponent, decorators: [{
            type: Component,
            args: [{
                    template: ''
                }]
        }] });
/*
  This is an unused module to resolve the ng build error:
    'Cannot determine the module for class HostComponent'

  Reference: https://github.com/angular/issues/13590
*/
export class HostModule {
}
HostModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: HostModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
HostModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: HostModule, declarations: [HostComponent] });
HostModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: HostModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.2", ngImport: i0, type: HostModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [HostComponent]
                }]
        }] });
