// hdr-histogram-js/src/index.ts
/*
 * This is a TypeScript port of the original Java version, which was written by
 * Gil Tene as described in
 * https://github.com/HdrHistogram/HdrHistogram
 * and released to the public domain, as explained at
 * http://creativecommons.org/publicdomain/zero/1.0/
 */
import ByteBuffer from "hdr-histogram-js/src/ByteBuffer";
import {
  decodeFromCompressedBase64,
  encodeIntoCompressedBase64,
} from "hdr-histogram-js/src/encoding";
import type Histogram from "hdr-histogram-js/src/Histogram";
import type { HistogramSummary, BitBucketSize } from "hdr-histogram-js/src/Histogram";
import Float64Histogram from "hdr-histogram-js/src/Float64Histogram";
import HistogramLogReader, { listTags } from "hdr-histogram-js/src/HistogramLogReader";
import HistogramLogWriter from "hdr-histogram-js/src/HistogramLogWriter";
import Int16Histogram from "hdr-histogram-js/src/Int16Histogram";
import Int32Histogram from "hdr-histogram-js/src/Int32Histogram";
import Int8Histogram from "hdr-histogram-js/src/Int8Histogram";
import JsHistogram from "./JsHistogram";
import PackedHistogram from "hdr-histogram-js/src/PackedHistogram";
import Recorder from "hdr-histogram-js/src/Recorder";
import { initWebAssembly, initWebAssemblySync, WasmHistogram } from "./wasm";
import type { BuildRequest } from "hdr-histogram-js/src/HistogramBuilder";
import { build } from "hdr-histogram-js/src/HistogramBuilder";

export {
  initWebAssembly,
  initWebAssemblySync,
  Histogram,
  BitBucketSize,
  HistogramSummary,
  Int8Histogram,
  Int16Histogram,
  Int32Histogram,
  Float64Histogram,
  PackedHistogram,
  HistogramLogReader,
  listTags,
  build,
  BuildRequest,
  ByteBuffer,
  decodeFromCompressedBase64,
  encodeIntoCompressedBase64,
  HistogramLogWriter,
  Recorder,
  WasmHistogram,
  JsHistogram,
};
