// hdr-histogram-js/src/JsHistogramFactory.ts
import { BitBucketSize } from "hdr-histogram-js/src/Histogram";
import PackedHistogram from "hdr-histogram-js/src/PackedHistogram";
import Int8Histogram from "hdr-histogram-js/src/Int8Histogram";
import Int16Histogram from "hdr-histogram-js/src/Int16Histogram";
import Int32Histogram from "hdr-histogram-js/src/Int32Histogram";
import Float64Histogram from "hdr-histogram-js/src/Float64Histogram";
import JsHistogram from "./JsHistogram";

export interface JsHistogramConstructor {
  new (
    lowestDiscernibleValue: number,
    highestTrackableValue: number,
    numberOfSignificantValueDigits: number
  ): JsHistogram;
}

export function constructorFromBucketSize(
  bitBucketSize: BitBucketSize
): JsHistogramConstructor {
  switch (bitBucketSize) {
    case "packed":
      return PackedHistogram;
    case 8:
      return Int8Histogram;
    case 16:
      return Int16Histogram;
    case 32:
      return Int32Histogram;
    case 64:
      return Float64Histogram;
    default:
      throw new Error("Incorrect parameter bitBucketSize");
  }
}
